src/main/java/com/powsybl/openloadflow/ac/PhaseControlOuterLoop.java

```diff
@@ -87,4 +91,38 @@ private void switchOffPhaseControl(DiscretePhaseControl phaseControl) {
         double roundedA1Value = piModel.getA1();
         LOGGER.info("Round phase shift of '{}': {} -> {}", controllerBranch.getId(), a1Value, roundedA1Value);
     }
+
+    private OuterLoopStatus changeTapPositions(OuterLoopContext context, DiscretePhaseControl phaseControl) {
+        // only local control supported: controlled branch is controller branch.
+        double currentLimit = phaseControl.getTargetValue();
+        LfBranch controllerBranch = phaseControl.getController();
+        PiModel piModel = controllerBranch.getPiModel();
+        boolean isSensibilityPositive;
+        boolean success = false;
+        if (phaseControl.getControlledSide() == DiscretePhaseControl.ControlledSide.ONE && currentLimit < controllerBranch.getI1().eval()) {
+            isSensibilityPositive = isSensitivityCurrentPerA1Positive(context.getVariableSet(), controllerBranch, DiscretePhaseControl.ControlledSide.ONE);
+            success = isSensibilityPositive ? piModel.getNewTapPosition(PiModel.Direction.DECREASE) : piModel.getNewTapPosition(PiModel.Direction.INCREASE);
+        } else if (phaseControl.getControlledSide() == DiscretePhaseControl.ControlledSide.TWO && currentLimit < controllerBranch.getI2().eval()) {
+            isSensibilityPositive = isSensitivityCurrentPerA1Positive(context.getVariableSet(), controllerBranch, DiscretePhaseControl.ControlledSide.TWO);
+            success = isSensibilityPositive ? piModel.getNewTapPosition(PiModel.Direction.DECREASE) : piModel.getNewTapPosition(PiModel.Direction.INCREASE);
+        }
+        if (success) {
+            for (LfNetworkListener listener : controllerBranch.getNetwork().getListeners()) {
```

Comment:
Once notification will be managed inside `PiModel` we won't need this hack anymore because the call to `getNewTapPosition` will trigger the notif.

Reply:
Done.